crates/pallet-transaction-fees/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+            /// Receiver of the tip.
+            who: T::AccountId,
+            /// Amount of collected tips.
+            amount: BalanceOf<T>,
+        },
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+    where
+        BalanceOf<T>: From<u8> + From<u64>,
+    {
+        fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+            Self::do_initialize(now);
+            T::WeightInfo::on_initialize()
+        }
+
+        fn on_finalize(now: BlockNumberFor<T>) {
+            Self::do_finalize(now);
+        }
+    }
+}
+
+impl<T: Config> Pallet<T>
+where
+    BalanceOf<T>: From<u64>,
+{
+    fn do_initialize(_n: T::BlockNumber) {
+        let block_author = T::FindAuthor::find_author(
+            frame_system::Pallet::<T>::digest()
+                .logs
+                .iter()
+                .filter_map(|d| d.as_pre_runtime()),
+        )
+        .expect("Block author must always be present; qed");
+
+        BlockAuthor::<T>::put(block_author);
+
+        CollectedBlockFees::<T>::put(CollectedFees {
+            storage: BalanceOf::<T>::zero(),
+            compute: BalanceOf::<T>::zero(),
+            tips: BalanceOf::<T>::zero(),
+        });
+    }
+
+    // TODO: Fees will be split between farmers and executors in the future
+    fn do_finalize(_n: T::BlockNumber) {
+        TransactionByteFee::<T>::take();
+
+        let block_author =
+            BlockAuthor::<T>::take().expect("`BlockAuthor` was set in `on_initialize`; qed");
+
+        let original_storage_fees_escrow = CollectedStorageFeesEscrow::<T>::get();
+        let collected_fees = CollectedBlockFees::<T>::take()
+            .expect("`CollectedBlockFees` was set in `on_initialize`; qed");
+
+        let mut storage_fees_escrow = original_storage_fees_escrow;
+
+        // Take a portion of storage fees escrow as a farmer reward.
+        let storage_fees_escrow_reward = storage_fees_escrow
+            / T::StorageFeesEscrowBlockReward::get().1.into()
+            * T::StorageFeesEscrowBlockReward::get().0.into();
+        storage_fees_escrow -= storage_fees_escrow_reward;
+
+        // Take a portion of storage fees collected in this block as a farmer reward.
+        let collected_storage_fees_reward = collected_fees.storage
+            / T::StorageFeesEscrowBlockTax::get().1.into()
+            * (T::StorageFeesEscrowBlockTax::get().1 - T::StorageFeesEscrowBlockTax::get().0)
+                .into();
+        storage_fees_escrow += collected_fees.storage - collected_storage_fees_reward;
+
+        // Update storage fees escrow.
+        if storage_fees_escrow != original_storage_fees_escrow {
+            CollectedStorageFeesEscrow::<T>::put(storage_fees_escrow);
+            Self::deposit_event(Event::<T>::StorageFeesEscrowChange {
+                before: original_storage_fees_escrow,
+                after: storage_fees_escrow,
+            });
+        }
+
+        // Issue storage fees reward.
```

Comment:
I'm pondering that whether it's better to have such an event:

```rust
enum Event {
    FeeRewards  { who: T::AccountId, details: RewardDetails }
}

struct RewardDetails<Balance> {
    storage: Option<Balance>,
    compute: Option<Balance>,
    tips: Option<Balance>,
}
```

Reply:
The idea there is that if we combine them then it will become irrelevant once we have decoupled execution because inevitably some of the fields will be zeroes: compute fees for farmers and storage fees for executors, at least with current implementation. Considering that tip is optional, in most cases these events will have all fields except one set to zero, in which case it makes more sense to have separate events.

We can change this later once we have more insides into what will actually happen and what frontend folks will find useful to have as events.

Reply:
Yeah, we could think over it later, just thinking that from the aesthetic perspective :P